util/postprocess-makedepend.pl

```diff
@@ -0,0 +1,95 @@
+        #     makedepend -f- -o"|dir/foo.o" -- $(CFLAGS) -- ../somewhere/foo.c
+        #
+        # The result will look something like this:
+        #
+        #     ../somewhere/foo|dir/foo.o: deps...
+        #
+        # Which is easy to massage by removing everything up to the first |
+
+        # Remove everything up to the first |
+        s/^.*\|//;
+        # Also, remove any dependency that starts with a /, because those are
+        # typically system headers
+        s/ \/(\\.|[^ ])*//;
+        # Finally, discard all empty lines or comment lines
+        $_ = undef if (/: *$/ || /^(#.*| *)$/);
+        $_.="\n" unless !defined($_) or /\R$/g;
+
+    } elsif ($producer eq 'VMS C') {
+
+        # current versions of DEC / Compaq / HP / VSI C strips away all
+        # directory information from the object file, so we must insert it
+        # back. Just to be safe against future changes, we check that there
+        # really is no directory information.
+
+        my $directory = $ARGV[1];
+
+        # The pattern for target and dependencies will always take this form:
+        #
+        #     target SPACE : SPACE deps
+        #
+        # This is so a volume delimiter (a : without any spaces around it)
+        # won't get mixed up with the target / deps delimiter.  We use this
+        # fact in the regexp below to make sure we do look at the target.
+        s/^/$directory/ unless /^\S+[:>\]]\S+\s+:/;
+
+        # We know that VMS has system header files in text libraries,
+        # extension .TLB.  We also know that our header files aren't stored
+        # in text libraries.  Finally, we know that VMS C produces exactly
+        # one dependency per line, so we simply discard any line ending with
+        # .TLB.
+        $_ = undef if /\.TLB\s*$/;
+
+    } elsif ($producer eq 'VC') {
+
+        # For the moment, we only know of one native Windows C compiler, and
+        # that's Visual C.  With that compiler, the flags /Zs /showIncludes
+        # give us the necessary output to be able to create dependencies that
+        # nmake (or any 'make' implementation) should be able to read, with a
+        # bit of help.  The output we're interested in looks like this (it
+        # always starts the same)
+        #
+        #     Note: including file: {whatever header file}
+        #
+        # So all we really have to do is to is to replace the start of the line
+        # with an object file specification, given to us as $ARGV[1].
+        #
+        # There are also other lines mixed in, for example compiler warnings,
+        # so we simply discard anything that doesn't start with the Note:
+
+        my $object = $ARGV[1];
+        if (/^Note: including file: */) {
+            $_ = "${object}: ".$';
+        } else {
+            $_ = undef;
+        }
+
+    } else {
+        if ($producer) {
+            die "Producer unrecognised: $producer\n";
+        } else {
+            die "Producer not given\n";
+        }
```

Comment:
This is just ugly. Checking for command line argument validity in the loop. I understand that the condition is met only once, but that's not how you'd expect program look like... You kind of validate input, then loop. Actually one can say the same about overall structure, in sense that you don't expect *that* kind of condition in the loop... Well, I want to say this is a side note, but just can't collect myself :-)

Reply:
You're right. I'll look into refactoring tomorrow...